app/routes/user/password.tsx

```diff
@@ -0,0 +1,276 @@
+/*!
+ * Copyright © 2022 United States Government as represented by the Administrator
+ * of the National Aeronautics and Space Administration. No copyright is claimed
+ * in the United States under Title 17, U.S. Code. All Other Rights Reserved.
+ *
+ * SPDX-License-Identifier: NASA-1.3
+ */
+import type { DataFunctionArgs } from '@remix-run/node'
+import { useFetcher } from '@remix-run/react'
+import {
+  Button,
+  ButtonGroup,
+  Icon,
+  InputGroup,
+  Label,
+  TextInput,
+} from '@trussworks/react-uswds'
+import { useEffect, useRef, useState } from 'react'
+
+import { getUser } from '../__auth/user.server'
+import { updatePassword } from './password.server'
+import Spinner from '~/components/Spinner'
+import { getFormDataString } from '~/lib/utils'
+import { useUserIdp } from '~/root'
+
+export const handle = {
+  breadcrumb: 'Reset Password',
+  getSitemapEntries: () => null,
+}
+
+export async function loader({ request }: DataFunctionArgs) {
+  const user = await getUser(request)
+  if (!user) throw new Response(null, { status: 403 })
+  return null
+}
+
+export async function action({ request }: DataFunctionArgs) {
+  const data = await request.formData()
+  const oldPassword = getFormDataString(data, 'oldPassword')
+  const newPassword = getFormDataString(data, 'newPassword')
+
+  let response = null
+  if (oldPassword && newPassword) {
+    response = await updatePassword(request, oldPassword, newPassword)
+  }
+
+  return response
+}
+
+export default function () {
+  const idp = useUserIdp()
+  if (idp)
+    throw new Error(
+      'you must be logged in with a user name and password to reest password'
+    )
+  const fetcher = useFetcher<typeof action>()
+  const errorMessage = fetcher.data
+  const checkPassword = (str: string) => {
+    var re =
+      /^(?=.*\d)(?=.*[~`!@#$%^&*.+=\-_ [\]\\';,/{}()|\\":<>?])(?=.*[a-z])(?=.*[A-Z]).{8,}$/
+    return re.test(str)
+  }
+  const checkWhitespace = (str: string) => {
+    return /^\s|\s$/.test(str)
+  }
+
+  const [newPassword, setNewPassword] = useState('')
+  const [isOldPasswordTouched, setIsOldPasswordTouched] = useState(false)
+  const [isNewPasswordTouched, setIsNewPasswordTouched] = useState(false)
+  const [visible, setVisible] = useState(false)
+
+  const containsLower = /[a-z]/.test(newPassword)
+  const containsUpper = /[A-Z]/.test(newPassword)
+  const containsNumber = /[0-9]/.test(newPassword)
+  const validLength = newPassword.length >= 8
+  const containsSpecialChar = /[~`!@#$%^.&*+=\-_ [\]\\';,/{}()|\\":<>?]/g.test(
+    newPassword
+  )
+  const leadingOrTrailingSpace =
+    newPassword.length !== newPassword.trim().length
+  const oldPasswordError = errorMessage && !isOldPasswordTouched
+  const valid = !checkWhitespace(newPassword) && checkPassword(newPassword)
```

Comment:
Please construct the value of `valid` as a boolean combination of all of the conditions that you display in the checklist so that we can remove the redundant regex in the `checkPassword` function.

That way, if we ever change the password requirements, we only have to change the code in _one_ place.